src/main/java/uk/gov/hmcts/ccd/domain/service/search/CaseSearchResultGenerator.java

```diff
@@ -0,0 +1,177 @@
+            searchResultField.isMetadata(),
+            displayContextParameter(searchResultField, commonField));
+    }
+
+    private boolean filterDistinctFieldsByRole(final HashSet<String> addedFields, final SearchResultField resultField) {
+        String id = resultField.buildCaseFieldId();
+        if (addedFields.contains(id)) {
+            return false;
+        } else {
+            if (StringUtils.isEmpty(resultField.getRole()) || userRepository.getUserRoles().contains(resultField.getRole())) {
+                addedFields.add(id);
+                return true;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    private CommonField commonField(SearchResultField searchResultField, CaseFieldDefinition caseFieldDefinition) {
+        return caseFieldDefinition.getComplexFieldNestedField(searchResultField.getCaseFieldPath())
+            .orElseThrow(() ->
+                new BadRequestException(format("CaseField %s has no nested elements with code %s.",
+                    caseFieldDefinition.getId(), searchResultField.getCaseFieldPath())));
+    }
+
+    private String displayContextParameter(SearchResultField searchResultField, CommonField commonField) {
+        return searchResultField.getDisplayContextParameter() == null
+            ? commonField.getDisplayContextParameter()
+            : searchResultField.getDisplayContextParameter();
+    }
+
+    private SearchResultViewItem buildSearchResultViewItem(final CaseDetails caseDetails,
+                                                           final CaseTypeDefinition caseTypeDefinition,
+                                                           final SearchResult searchResult) {
+
+        Map<String, JsonNode> caseData = new HashMap<>(caseDetails.getData());
+        Map<String, Object> caseMetadata = new HashMap<>(caseDetails.getMetadata());
+        Map<String, TextNode> labels = caseTypeDefinition.getLabelsFromCaseFields();
+        Map<String, Object> caseFields = prepareData(searchResult, caseData, caseMetadata, labels);
+
+        String caseId = caseDetails.hasCaseReference() ? caseDetails.getReferenceAsString() : caseDetails.getId();
+        return new SearchResultViewItem(caseId, caseFields, new HashMap<>(caseFields));
+    }
+
+    private Map<String, Object> prepareData(SearchResult searchResult,
+                                            Map<String, JsonNode> caseData,
+                                            Map<String, Object> metadata,
+                                            Map<String, TextNode> labels) {
+
+        Map<String, Object> newResults = new HashMap<>();
+
+        searchResult.getFieldsWithPaths().forEach(searchResultField -> {
+            JsonNode jsonNode = caseData.get(searchResultField.getCaseFieldId());
+            if (jsonNode != null) {
+                newResults.put(searchResultField.getCaseFieldId() + "." + searchResultField.getCaseFieldPath(),
```

Comment:
I can see this is duplicated in a couple of places. Can you add a method to the searchResultField class called getIdIncludingPath() or sim?